lib/gamnit/virtual_gamepad/virtual_gamepad.nit

```diff
@@ -0,0 +1,418 @@
+
+# Control composing a `VirtualGamepad`
+abstract class RoundControl
+	# Center position on the UI
+	var center: Point3d[Float]
+
+	# Radius in UI units/pixels of the  part of this control
+	fun radius: Float is abstract
+
+	private fun sprites: Array[Sprite] do return new Array[Sprite]
+
+	private fun accept_event(event: InputEvent, ui_pos: Point[Float]): Bool
+	do
+		if event isa PointerEvent then
+			if contains(ui_pos) then
+				var hit = hit(event, ui_pos)
+				if hit then event
+				return hit
+			end
+		end
+
+		return false
+	end
+
+	# Does `self` contain a pointer at `ui_pos`?
+	private fun contains(ui_pos: Point[Float]): Bool
+	do
+		var dx = center.x - ui_pos.x
+		var dy = center.y - ui_pos.y
+		return dx*dx + dy*dy < radius*radius
+	end
+
+	private fun hit(event: PointerEvent, ui_pos: Point[Float]): Bool
+	do return false
+
+	# Keys currently down, to be depressed if the pointer moves away
+	private var down_names = new Set[String]
+
+	# Depress/release keys kept down, listed by `down_names`
+	private fun depressed_down
+	do
+		for name in down_names do
+			var e = new VirtualGamepadEvent(name)
+			e.is_down = false
+			app.accept_event e
+		end
+		down_names.clear
+	end
+end
+
+# Simple action button
+class RoundButton
+	super RoundControl
+
+	# Event name, should usually correspond to a keyboard key like "a" or "left"
+	var name: String
+
+	# Texture drawn for this button, may be from `app.gamepad_spritesheet`
+	var texture: Texture
+
+	redef fun radius do return 0.5*texture.height
+
+	redef fun hit(event, ui_pos)
+	do
+		if not event.is_move then
+			var e = new VirtualGamepadEvent(name.to_s)
+			e.is_down = event.pressed
+			app.accept_event e
+
+			if event.pressed then
+				down_names.add name
+			else down_names.clear
+		end
+		return true
+	end
+
+	redef var sprites = [new Sprite(texture, center)] is lazy
+end
+
+# Directional pad with up to 4 buttons
+#
+# Assumes that each pad is manipulated by at max a single pointer.
+class DPad
+	super RoundControl
+
+	# Event names for the keys, in order of top, left, down and right
+	var names: Array[String]
+
+	# Show the up button
+	var show_up = true is writable
+
+	# Show the down button
+	var show_down = true is writable
+
+	# Show the left button
+	var show_left = true is writable
+
+	# Show the right button
+	var show_right = true is writable
+
+	redef fun radius do return 200.0
+
+	redef fun contains(ui_pos)
+	do
+		if show_down then return super(new Point[Float](ui_pos.x+0.0, ui_pos.y-100.0))
+		return super
+	end
+
+	redef fun hit(event, ui_pos)
+	do
+		var display = app.display
+		if display == null then return false
+
+		var dx = ui_pos.x - center.x
+		var dy = ui_pos.y - center.y
+		if show_down then dy -= 100.0
+
+		# Angle (with 0.0 on the right) to index in WASD (0 -> W/up)
+		var indexes = new Set[Int]
+		var ao = atan2(dy, dx)
+		ao -= pi/4.0
+
+		# Look for 2 angles so 2 buttons can be pressed at the same time
+		for da in once [-pi/8.0, pi/8.0] do
+			var a = ao+da
```

Comment:
Looks like @privat variable naming haha, but honestly I don't see good names for thoses